In [762]:
# import the nexessary libraries for the data extraction and analysis
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

import statsmodels.api as sm
from scipy import stats
import seaborn as sns

In [763]:
# read the csv file
df = pd.read_csv('mosquito_data_part_2.csv')

In [764]:
# take a look at the data
df.sample(10)

Year  Week          Address Block  Trap Trap type                 Date  \
12054  2009    30      5XX S CENTRAL AVE  T031    GRAVID  2009-07-31 00:07:00   
15375  2007    23      40XX N AUSTIN AVE  T142    GRAVID  2007-06-29 11:06:17   
8425   2012    33       127XX S DOTY AVE  T115    GRAVID  2012-08-17 00:08:00   
16987  2016    30     14XX N HUMBOLDT DR  T033    GRAVID  2016-07-27 00:07:00   
11472  2010    25  100XX W OHARE AIRPORT  T912    GRAVID  2010-06-28 00:06:00   
13777  2007    39      37XX S PULASKI RD  T063    GRAVID  2007-10-04 00:10:00   
1048   2019    24   62XX N MCCLELLAN AVE  T236    GRAVID  2019-06-13 00:06:00   
4506   2015    31     25XX S MILLARD AVE  T152    GRAVID  2015-08-06 00:08:00   
1246   2018    34    89XX S CARPENTER ST  T159    GRAVID  2018-08-23 00:08:00   
9142   2012    24     82XX S KOSTNER AVE  T225    GRAVID  2012-06-15 00:06:00   

       Mosquito number WNV Present           Species        Lat        Lon  \
12054                5    negative     CULEX PIPIENS  41.872873 -87.764737   
15375                1    negative    CULEX RESTUANS  41.953540 -87.776939   
8425                11    negative    CULEX RESTUANS  41.662387 -87.590180   
16987               15    positive    CULEX RESTUANS  41.907288 -87.701391   
11472               50    negative    CULEX RESTUANS  41.977738 -87.880969   
13777                2    negative     CULEX PIPIENS  41.825398 -87.723962   
1048                 1    negative   CULEX TERRITANS  41.994966 -87.770837   
4506                 2    negative    CULEX RESTUANS  41.845715 -87.715899   
1246                 1    negative  CULEX SALINARIUS  41.731447 -87.649722   
9142                 2    negative    CULEX RESTUANS  41.742670 -87.731550   

       Month  
12054      7  
15375      6  
8425       8  
16987      7  
11472      6  
13777     10  
1048       6  
4506       8  
1246       8  
9142       6

In [765]:
#learn about the data info like the columns and datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18495 entries, 0 to 18494
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             18495 non-null  int64  
 1   Week             18495 non-null  int64  
 2   Address Block    18495 non-null  object 
 3   Trap             18495 non-null  object 
 4   Trap type        18495 non-null  object 
 5   Date             18495 non-null  object 
 6   Mosquito number  18495 non-null  int64  
 7   WNV Present      18495 non-null  object 
 8   Species          18495 non-null  object 
 9   Lat              18495 non-null  float64
 10  Lon              18495 non-null  float64
 11  Month            18495 non-null  int64  
dtypes: float64(2), int64(4), object(6)
memory usage: 1.7+ MB


# Part-1.1 Convert the WNV Present column into a binary column and create dummy variables from the Trap type column.

In [766]:
df['WNV Present'] = df['WNV Present'].map({'positive':1,'negative':0,})

We assign positive as 1 and negative as 2

In [767]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18495 entries, 0 to 18494
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             18495 non-null  int64  
 1   Week             18495 non-null  int64  
 2   Address Block    18495 non-null  object 
 3   Trap             18495 non-null  object 
 4   Trap type        18495 non-null  object 
 5   Date             18495 non-null  object 
 6   Mosquito number  18495 non-null  int64  
 7   WNV Present      18495 non-null  int64  
 8   Species          18495 non-null  object 
 9   Lat              18495 non-null  float64
 10  Lon              18495 non-null  float64
 11  Month            18495 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 1.7+ MB


In [768]:
df['WNV Present']

0        0
1        0
2        0
3        0
4        0
        ..
18490    1
18491    1
18492    1
18493    1
18494    1
Name: WNV Present, Length: 18495, dtype: int64

In [769]:
#By using info(), we now can see that data type is changed to integer.
df['WNV Present'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 18495 entries, 0 to 18494
Series name: WNV Present
Non-Null Count  Dtype
--------------  -----
18495 non-null  int64
dtypes: int64(1)
memory usage: 144.6 KB


In [770]:
#By using pd.get_dummies(), we can create dummy variables.
pd.get_dummies(df['Trap type'])

CDC  GRAVID  OVI  SENTINEL
0        0       1    0         0
1        0       1    0         0
2        0       1    0         0
3        0       1    0         0
4        0       1    0         0
...    ...     ...  ...       ...
18490    0       1    0         0
18491    0       1    0         0
18492    0       1    0         0
18493    0       0    0         1
18494    0       1    0         0

[18495 rows x 4 columns]

# Part 1.2: What is the average number of mosquitoes for each month? What trends do you notice?

In [771]:
df.groupby('Month').sum()[['Mosquito number']]

C:\Users\Jason\AppData\Local\Temp\ipykernel_9780\1141903469.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby('Month').sum()[['Mosquito number']]


Mosquito number
Month                 
5                  116
6                28607
7                59007
8                83229
9                29055
10                1210

Mosquito number peaks in July and August and falls starting from Spetember. The decreasing tends continues until May, and increases afterwards. We can tell that the numbe of mosquitoes increases in summerttime and decreases in winter time. 

# Part 2.1. Is there a statistically significant difference between the different mosquito species when looking at the occurrence of West Nile Virus?

In [772]:
# First, take a look at the species 
df["Species"].unique()

array(['CULEX RESTUANS', 'CULEX TERRITANS', 'CULEX SALINARIUS',
       'CULEX PIPIENS'], dtype=object)

In [773]:
# Perfrom Anova analyses to measure the statistical analyses
anova_data = {}
mosquito_types = df["Species"].unique()

for mosquito in mosquito_types:
    anova_data[mosquito] = df.loc[df["Species"] == mosquito, "WNV Present"]

In [774]:
stats.f_oneway(anova_data["CULEX RESTUANS"], 
               anova_data["CULEX TERRITANS"], 
               anova_data["CULEX SALINARIUS"],
               anova_data["CULEX PIPIENS"])

F_onewayResult(statistic=105.4527050388844, pvalue=1.082615440825039e-67)

Since the pvalue is p-value is <0.05, we can reject the null hypothesis (H0 says there is no difference), and conclude that there is a statistically significant difference between the different mosquito species when looking at the occurrence of West Nile Virus.

# Part 2.2 Which columns are positively correlated with the number of mosquitoes caught? Which columns are negatively correlated? Are these correlations statistically significant?

In [775]:
CM = df.corr(numeric_only=True)
CM_mosquito = CM['Mosquito number'].sort_values(ascending=False)
print(CM_mosquito)

Mosquito number    1.000000
WNV Present        0.408034
Year               0.129326
Lat                0.096820
Week              -0.033900
Month             -0.040426
Lon               -0.151421
Name: Mosquito number, dtype: float64


Columns that are positively correlated with the number of mosquitoes caught inlcude: Year, WNV Present, Lat. Columns that are negatively correlated with the number of mosquitoes caught inlcude: Week, Lon, and month. 

In [776]:
# Calculate correlation coefficient and p-value
correlation_coefficient, p_value = stats.pearsonr(df['Mosquito number'], df['Year'])

# Print correlation coefficient and p-value
print("Correlation Coefficient:", correlation_coefficient)
print("P-value:", p_value)

Correlation Coefficient: 0.12932594549371043
P-value: 8.393050016377127e-70


p< 0.05; the correlation is statistically significant

In [777]:
# Calculate correlation coefficient and p-value
correlation_coefficient, p_value = stats.pearsonr(df['Mosquito number'], df['Week'])

# Print correlation coefficient and p-value
print("Correlation Coefficient:", correlation_coefficient)
print("P-value:", p_value)

Correlation Coefficient: -0.0338996075449608
P-value: 4.0028248035811835e-06


p is smaller than < 0.05; the correlation is statistically significant

In [778]:
# Calculate correlation coefficient and p-value
correlation_coefficient, p_value = stats.pearsonr(df['Mosquito number'], df['Lat'])

# Print correlation coefficient and p-value
print("Correlation Coefficient:", correlation_coefficient)
print("P-value:", p_value)

Correlation Coefficient: 0.09681989674648903
P-value: 9.096861337562708e-40


pvalue is smaller than 0.05; the correlation is  statistically significant

In [779]:
# Calculate correlation coefficient and p-value
correlation_coefficient, p_value = stats.pearsonr(df['Mosquito number'], df['Lon'])

# Print correlation coefficient and p-value
print("Correlation Coefficient:", correlation_coefficient)
print("P-value:", p_value)

Correlation Coefficient: -0.151421125968909
P-value: 2.7656039821185354e-95


pvalue is smaller than 0.05; the correlation is statistically significant

In [780]:
# Calculate correlation coefficient and p-value
correlation_coefficient, p_value = stats.pearsonr(df['Mosquito number'], df['Month'])

# Print correlation coefficient and p-value
print("Correlation Coefficient:", correlation_coefficient)
print("P-value:", p_value)

Correlation Coefficient: -0.040426437196600794
P-value: 3.8038755209747194e-08


p value is smaller 0.05; the correlation is statistically significant

# Part 3-1 Run a linear regression to determine how the independent variables affect the number of mosquitoes caught. Explain your model construction process. Analyze the model and the results and discuss the model’s limitations. This may end up being an iterative process.


In [754]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18495 entries, 0 to 18494
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             18495 non-null  int64  
 1   Week             18495 non-null  int64  
 2   Address Block    18495 non-null  object 
 3   Trap             18495 non-null  object 
 4   Trap type        18495 non-null  object 
 5   Date             18495 non-null  object 
 6   Mosquito number  18495 non-null  int64  
 7   WNV Present      18495 non-null  int64  
 8   Species          18495 non-null  object 
 9   Lat              18495 non-null  float64
 10  Lon              18495 non-null  float64
 11  Month            18495 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 1.7+ MB


In [755]:
# There are too many columns, drop them
df = df.drop(['Address Block', 'Date', 'WNV Present', 'Trap'], axis=1)

In [756]:
df.head()

Year  Week Trap type  Mosquito number         Species        Lat  \
0  2019    39    GRAVID                2  CULEX RESTUANS  41.977738   
1  2019    39    GRAVID                1  CULEX RESTUANS  41.798211   
2  2019    39    GRAVID                2  CULEX RESTUANS  41.987245   
3  2019    39    GRAVID                1  CULEX RESTUANS  41.953664   
4  2019    39    GRAVID                9  CULEX RESTUANS  41.657069   

         Lon  Month  
0 -87.880969      9  
1 -87.736925      9  
2 -87.689417      9  
3 -87.724987      9  
4 -87.546049      9

In [757]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18495 entries, 0 to 18494
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             18495 non-null  int64  
 1   Week             18495 non-null  int64  
 2   Trap type        18495 non-null  object 
 3   Mosquito number  18495 non-null  int64  
 4   Species          18495 non-null  object 
 5   Lat              18495 non-null  float64
 6   Lon              18495 non-null  float64
 7   Month            18495 non-null  int64  
dtypes: float64(2), int64(4), object(2)
memory usage: 1.1+ MB


In [758]:
# Create dummy variables for Trap type and species
df_encoded = pd.get_dummies(df, columns=['Trap type', 'Species'], drop_first=True)

In [759]:
df_encoded.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18495 entries, 0 to 18494
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Year                      18495 non-null  int64  
 1   Week                      18495 non-null  int64  
 2   Mosquito number           18495 non-null  int64  
 3   Lat                       18495 non-null  float64
 4   Lon                       18495 non-null  float64
 5   Month                     18495 non-null  int64  
 6   Trap type_GRAVID          18495 non-null  uint8  
 7   Trap type_OVI             18495 non-null  uint8  
 8   Trap type_SENTINEL        18495 non-null  uint8  
 9   Species_CULEX RESTUANS    18495 non-null  uint8  
 10  Species_CULEX SALINARIUS  18495 non-null  uint8  
 11  Species_CULEX TERRITANS   18495 non-null  uint8  
dtypes: float64(2), int64(4), uint8(6)
memory usage: 975.4 KB


In [760]:
X = df_encoded.drop('Mosquito number', axis=1)  # Independent variables (features)
y = df_encoded['Mosquito number']  # Dependent variable (target)

In [761]:
# Add a constant column to X
X = sm.add_constant(X)

# Fit the linear regression model
model = sm.OLS(y, X)
lm = model.fit()

# Print the summary of the model
print(lm.summary())

                            OLS Regression Results                            
Dep. Variable:        Mosquito number   R-squared:                       0.093
Model:                            OLS   Adj. R-squared:                  0.092
Method:                 Least Squares   F-statistic:                     171.3
Date:                Sat, 10 Jun 2023   Prob (F-statistic):               0.00
Time:                        22:11:15   Log-Likelihood:                -73448.
No. Observations:               18495   AIC:                         1.469e+05
Df Residuals:                   18483   BIC:                         1.470e+05
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
const                   

R-squared : 
- The R-squared value is 0.093, indicating that approximately 9.3% of the variance in the dependent variable (Mosquito number) is explained by the independent variables in the model. 

Coefficients: 

- The coefficient for the "Year" variable is 0.5333, indicating that a one-unit increase in the "Year" variable is associated with a 0.5333 unit increase in the "Mosquito number" 
- he coefficient for the "Trap type_SENTINEL" variable is 2.0630, indicating that a one-unit increase in the "Trap type_SENTINEL" variable is associated with a 2.0630 unit increase in the "Mosquito number" 


p-value: 
- The p-value represents the probability of observing the coefficient estimate as extreme as the one calculated, assuming the null hypothesis is true (null hypothesis: the coefficient is equal to zero). - A low p-value (typically less than 0.05) suggests that the variable has a significant effect on the dependent variable.

# Part 3-2 Run a logistic regression to determine how the independent variables affect West Nile Virus presence. Explain your model construction process. Analyze the model and the results and discuss the model’s limitations. This may end up being an iterative process.


In [781]:
df.dtypes

Year                 int64
Week                 int64
Address Block       object
Trap                object
Trap type           object
Date                object
Mosquito number      int64
WNV Present          int64
Species             object
Lat                float64
Lon                float64
Month                int64
dtype: object

In [782]:
df = df.drop(['Address Block', 'Date', 'Trap'], axis=1)

In [783]:
df_encoded = pd.get_dummies(df, columns=['Trap type', 'Species'], drop_first=True)

In [784]:
# Define my X and y
X = df_encoded.drop('WNV Present', axis=1)
y = df_encoded['WNV Present']

# adding the column of ones
X_withconstant = sm.add_constant(X)

# Fitting the model 

# 1. Instantiate model
WNVpresent= sm.Logit(y,X_withconstant)

# 2. Fit model
WNVpresent_results = WNVpresent.fit()

# Look at summary
WNVpresent_results.summary()

         Current function value: 0.405443
         Iterations: 35


C:\Users\Jason\anaconda3\lib\site-packages\statsmodels\base\model.py:604: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:            WNV Present   No. Observations:                18495
Model:                          Logit   Df Residuals:                    18482
Method:                           MLE   Df Model:                           12
Date:                Sat, 10 Jun 2023   Pseudo R-squ.:                  0.2229
Time:                        22:11:34   Log-Likelihood:                -7498.7
converged:                      False   LL-Null:                       -9649.5
Covariance Type:            nonrobust   LLR p-value:                     0.000
============================================================================================
                               coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------
const                     -517.7750     25.185    -20.559      0.000    -567.137    -468.412
Year                         0.1130      0.007     16.467      0.000       0.100       0.126
Week                         0.1002      0.016      6.243      0.000       0.069       0.132
Mosquito number              0.0651      0.001     44.694      0.000       0.062       0.068
Lat                         -0.6595      0.287     -2.301      0.021      -1.221      -0.098
Lon                         -3.5432      0.335    -10.580      0.000      -4.200      -2.887
Month                        0.2202      0.069      3.192      0.001       0.085       0.355
Trap type_GRAVID             0.1557      0.132      1.184      0.237      -0.102       0.413
Trap type_OVI              -13.7967   6889.959     -0.002      0.998   -1.35e+04    1.35e+04
Trap type_SENTINEL          -0.2383      0.215     -1.111      0.267      -0.659       0.182
Species_CULEX RESTUANS      -0.2502      0.051     -4.876      0.000      -0.351      -0.150
Species_CULEX SALINARIUS    -0.9952      0.322     -3.088      0.002      -1.627      -0.364
Species_CULEX TERRITANS     -2.4900      0.274     -9.072      0.000      -3.028      -1.952
============================================================================================
"""

The results of the logistic regression model are as follows:

The coefficients for the independent variables are as follows:

- Year: 0.1130, suggesting that with each unit increase in the year, the log-odds of West Nile Virus presence increase by 0.1130.
- Week: 0.1002, indicating that with each unit increase in the week, the log-odds of West Nile Virus presence increase by 0.1002.
- Mosquito number: 0.0651, suggesting that for each additional mosquito, the log-odds of West Nile Virus presence increase by 0.0651.
- Lat: -0.6595, indicating that with each unit increase in latitude, the log-odds of West Nile Virus presence decrease by 0.6595.
- Lon: -3.5432, suggesting that with each unit increase in longitude, the log-odds of West Nile Virus presence decrease by 3.5432.
- Month: 0.2202, indicating that for each month increase, the log-odds of West Nile Virus presence increase by 0.2202.
- Trap type_GRAVID: 0.1557, suggesting that the presence of the 'GRAVID' trap type increases the log-odds of West Nile Virus presence by 0.1557 compared to the reference trap type.
- Trap type_OVI: -13.7967, but it seems that the coefficient is extremely large and likely due to numerical instability or data issue. It may not be interpretable or meaningful.
- Trap type_SENTINEL: -0.2383, indicating that the presence of the 'SENTINEL' trap type decreases the log-odds of West Nile Virus presence by 0.2383 compared to the reference trap type.
- Species_CULEX RESTUANS: -0.2502, suggesting that the presence of the 'CULEX RESTUANS' species decreases the log-odds of West Nile Virus presence by 0.2502 compared to the reference species.
- Species_CULEX SALINARIUS: -0.9952, indicating that the presence of the 'CULEX SALINARIUS' species - decreases the log-odds of West Nile Virus presence by 0.9952 compared to the reference species.
- Species_CULEX TERRITANS: -2.4900, suggesting that the presence of the 'CULEX TERRITANS' species decreases the log-odds of West Nile Virus presence by 2.4900 compared to the reference species.

p-vale analyses:
-  'Year', 'Week', 'Mosquito number', 'Lat', 'Lon', 'Month', 'Species_CULEX RESTUANS', and 'Species_CULEX SALINARIUS' have p-values less than 0.05 and are considered statistically significant.

The limitations of the model may include:

- There may be other factors promoting the West Nile Virus presence including environmental factors like weather and the number of intermediate hosts of WNV like animals.